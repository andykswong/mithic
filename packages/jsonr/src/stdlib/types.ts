import { symFn } from '../symbol.ts';
import { AsyncValue, Env, FnValue, ObjValue } from '../types.ts';
import { assertFunction, isFunction, isObj, isString } from '../utils.ts';

/** Creates an array from elements. */
export function array(...vals: AsyncValue[]): AsyncValue[] {
  return vals;
}

/** Creates an object from entries. */
export function object(...entries: AsyncValue[]): ObjValue {
  const result = Object.create(null);
  for (let i = 0; i < entries.length; i += 2) {
    result[entries[i] as string] = entries[i + 1] ?? null;
  }
  return result;
}

/** Stringify given value, while sanitizing native objects and functions. */
export function string(val: AsyncValue): string {
  if (isFunction(val) && !val[symFn]) { return val.name || 'fn#anonymous'; }
  if (val && !Array.isArray(val) && !isObj(val) && typeof val === 'object') { return '{}'; }
  return `${val ?? null}`;
}

/** Returns given value as boolean */
export function boolean(a: AsyncValue): boolean {
  return !!a;
}

/** Converts a string to an integer number. */
export const parseInt = global.parseInt as FnValue;

/** Converts a string to a number. */
export const parseFloat = global.parseFloat as FnValue;

/** Merges arrays or objects. */
export function concat(val: AsyncValue, ...vals: AsyncValue[]): AsyncValue {
  if (!vals.length) { return val; }
  if (isObj(val)) { return Object.assign(Object.create(null), val, ...vals); }
  return ([] as AsyncValue[]).concat(val, ...vals);
}

/** Returns a subarray or substring. */
export function slice(val: AsyncValue, start: AsyncValue, end: AsyncValue): AsyncValue {
  if (Array.isArray(val)) { return val.slice(start as number, end as number); }
  if (isString(val)) { return val.slice(start as number, end as number); }
  return val;
}

/** Maps an array or value. */
export function map(this: Env, fn: AsyncValue, val: AsyncValue): AsyncValue {
  assertFunction(fn, 'arg0 of map');
  return Array.isArray(val) ? val.map(fn, this) : fn.call(this, val ?? null, 0, null);
}

/** Returns length of an array, string or object. */
export function length(iterable: AsyncValue): number {
  if (Array.isArray(iterable) || isString(iterable)) { return iterable.length; }
  if (isObj(iterable)) { return Object.keys(iterable).length; }
  return iterable === null ? 0 : 1;
}

/** Returns keys of an array, string or object. */
export function keys(iterable: AsyncValue): (string | number)[] {
  if (Array.isArray(iterable)) { return [...iterable.keys()]; }
  if (isString(iterable)) { return Array.from({ length: iterable.length }, (_, k) => k); }
  if (isObj(iterable)) { return Object.keys(iterable); }
  return [];
}

/** Returns values of an array, string or object. */
export function values(iterable: AsyncValue): AsyncValue[] {
  if (Array.isArray(iterable)) { return iterable; }
  if (isString(iterable)) { return [...iterable]; }
  if (isObj(iterable)) { return Object.values(iterable); }
  return [];
}

/** Returns entries of an array, string or object. */
export function entries(iterable: AsyncValue): [string | number, AsyncValue][] {
  if (Array.isArray(iterable)) { return [...iterable.entries()]; }
  if (isString(iterable)) { return [...[...iterable].entries()]; }
  if (isObj(iterable)) { return Object.entries(iterable); }
  return [];
}
