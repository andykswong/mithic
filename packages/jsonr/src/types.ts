import { MaybePromise } from '@mithic/commons';
import { symAsync, symEnv, symFn, symFnArgs, symFnBody, symMacro } from './symbol.ts';

/** metael AST evaluator. */
export interface Evaluator {
  /** Compiles given JSON AST into an unscoped function. */
  compile<Args extends unknown[] = AsyncValue[]>(expr: JSONType, args?: string[]): FnValue<Args>;

  /** Evaluates a JSON AST in given environment. Returns Promise for async code. */
  eval(expr: JSONType, env: Env): AsyncValue<Value>;
}

/** metael expression parser. */
export interface Parser {
  /** Parses given expression string into JSON AST. */
  parse(expr: string): JSONType;

  /** Prints given value or AST as string. */
  print(expr: AsyncValue): string;
}

/** Environment that defines variable scope. */
export interface Env {
  /** Returns the parent env scope, or null if the env is the global scope. */
  readonly parent: Env | null;

  /** Returns the global env scope. */
  readonly global: Env;

  /** Indicates if current scope is within a function. */
  fn: boolean;

  /** Indicates if the env is async. */
  async: boolean;

  /** Gets a variable. */
  get(name: string): AsyncValue | undefined;

  /** Gets a variable defined on current scope. */
  getOwn(name: string): AsyncValue | undefined;

  /** Sets a variable. */
  set<V extends AsyncValue>(name: string, value: V): V;

  /** Sets a variable on current scope. */
  setOwn<V extends AsyncValue>(name: string, value: V, readOnly?: boolean): V;

  /** Push a new scope to the stack and returns the new env. */
  push(): Env;
}

/** Variable bindings. */
export interface Bindings {
  [key: string | symbol]: AsyncValue;
}

/** Primitive types. */
export type Primitive = string | number | boolean;

/** Input value type. */
export type JSONType = null | Primitive | JSONType[] | { [key: string]: JSONType };

/** Awaited value type. */
export type Value =
  null | Primitive |
  AsyncValue[] |
  ObjValue |
  FnValue<AsyncValue[]>
  /** | NativeObject */;

/** A {@link Value} that may be wrapped with a Promise. */
export type AsyncValue<V = Value> = MaybePromise<V>;

/** Object value type. */
export type ObjValue = { [key: string]: AsyncValue };

/** Function value type. */
export type FnValue<Args extends unknown[] = AsyncValue[]>
  = ((this: Env, ...args: Args) => AsyncValue) & UserFnDef & AsyncFnDef & MacroFnDef;

/** Async function value type. */
export type AsyncFnValue<Args extends unknown[] = AsyncValue[]> = FnValue<Args> & Required<AsyncFnDef>;

/** Macro function value type. */
export type MacroValue<Args extends unknown[] = AsyncValue[]> = FnValue<Args> & Required<MacroFnDef>;

/** Macro function discriminator. */
export interface MacroFnDef {
  /** Discriminator for macros. */
  [symMacro]?: true;
}

/** Async function discriminator. */
export interface AsyncFnDef {
  /** Discriminator for async function. */
  [symAsync]?: true;
}

/** User function definitions. */
export interface UserFnDef {
  /** Discriminator for user defined functions. */
  [symFn]?: true;

  /** Env scope in which the function is declared in, or null for global scope. */
  [symEnv]?: Env | null;

  /** Argument names for this function. */
  [symFnArgs]?: string[];

  /** Function body. */
  [symFnBody]?: AsyncValue;
}
