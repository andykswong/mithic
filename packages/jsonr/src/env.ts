import { symAsync, symFn } from './symbol.ts';
import { AsyncValue, Bindings, Env } from './types.ts';

/** Environment that resolves values using object prototype chain. */
export class ProtoChainEnv implements Env {
  protected readonly binds: Bindings;

  public constructor(
    public readonly parent: ProtoChainEnv | null = null,
    constants: Bindings = {},
  ) {
    this.binds = Object.create(parent?.binds ?? null);
    const descriptors: PropertyDescriptorMap = {};
    for (const [key, value] of Object.entries(constants)) {
      descriptors[key] = { value, writable: false };
    }
    Object.defineProperties(this.binds, descriptors);
  }

  public get global(): ProtoChainEnv {
    return this.parent?.global ?? this;
  }

  public get fn(): boolean {
    return !!this.binds[symFn];
  }

  public set fn(isFn: boolean) {
    this.binds[symFn] = isFn;
  }

  public get async(): boolean {
    if (this.global === this) { return true; } // global scope is always async
    return !!this.binds[symAsync];
  }

  public set async(isAsync: boolean) {
    this.binds[symAsync] = isAsync;
  }

  public get(name: string): AsyncValue | undefined {
    return this.binds[name];
  }

  public getOwn(name: string): AsyncValue | undefined {
    return Object.hasOwn(this.binds, name) ? this.binds[name] : void 0;
  }

  public set<V extends AsyncValue>(name: string, value: V): V {
    if (!(name in this.binds)) {
      return (this.binds[name] = value) as V;
    }

    let binds = this.binds;
    while (!Object.hasOwn(binds, name)) { binds = Object.getPrototypeOf(binds); }
    return (binds[name] = value) as V;
  }

  public setOwn<V extends AsyncValue>(name: string, value: V, readOnly = false): V {
    if (readOnly) {
      Object.defineProperty(this.binds, name, { value, writable: false });
      return value;
    }
    return (this.binds[name] = value) as V;
  }

  public push(): ProtoChainEnv {
    return new ProtoChainEnv(this);
  }
}

/** Default environment implementation. */
export const DefaultEnv = ProtoChainEnv;
