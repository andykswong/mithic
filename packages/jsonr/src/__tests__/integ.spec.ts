import { beforeEach, describe, expect, it, jest } from '@jest/globals';
import { DefaultEnv } from '../env.ts';
import { Interpreter } from '../interpreter/index.ts';
import { JsonAstParser } from '../parser/index.ts';
import { Stdlib } from '../stdlib/index.ts';
import { AsyncValue, Env, Evaluator, JSONType, Parser } from '../types.ts';

import asyncTest from './data/async.json';
import macroTest from './data/macro.json';

describe('Integration tests', () => {
  let env: Env;
  let evaluator: Evaluator;
  let parser: Parser;
  let lines: AsyncValue[][];
  let readline: jest.Mock<(prompt: string, timeout?: number) => Promise<string>>;

  beforeEach(() => {
    env = new DefaultEnv(null, {
      ...Stdlib,
      println: (...args: AsyncValue[]) => (lines.push(args), null),
      readline: (...args: AsyncValue[]) => readline(...args as [string, number?]),
    });
    evaluator = new Interpreter();
    parser = new JsonAstParser();
    lines = [];
    readline = jest.fn();
  });

  describe('macro', () => {
    it('should print the correct results', async () => {
      const code = JSON.stringify(macroTest);
      const ast = parser.parse(code);
      const result = await evaluator.eval(ast, env);
      expect(result).toBe(null);
      expect(lines).toMatchSnapshot();
    });
  });

  describe('async', () => {
    let ast: JSONType;

    beforeEach(() => {
      const code = JSON.stringify(asyncTest);
      ast = parser.parse(code);
    });

    it('should print the correct results', async () => {
      const input = 'testing';
      readline.mockResolvedValueOnce(input);
      const result = await evaluator.eval(ast, env);
      expect(result).toBe(null);
      expect(lines).toMatchSnapshot();
      expect(readline).toHaveBeenCalledWith('your input: ', 2000);
    });

    it('should catch rejected promise from readline', async () => {
      readline.mockRejectedValueOnce(new Error('aborted'));
      const result = await evaluator.eval(ast, env);
      expect(result).toBe(null);
      expect(lines).toMatchSnapshot();
    });
  });
});
