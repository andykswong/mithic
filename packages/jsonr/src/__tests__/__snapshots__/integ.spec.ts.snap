// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration tests async should catch rejected promise from readline 1`] = `
[
  [
    "you said:",
    "<nope, too slow>",
  ],
]
`;

exports[`Integration tests async should print the correct results 1`] = `
[
  [
    "you said:",
    "testing",
  ],
]
`;

exports[`Integration tests macro should print the correct results 1`] = `
[
  [
    2,
  ],
  [
    3,
  ],
  [
    [
      "+",
      1,
      2,
    ],
  ],
]
`;
